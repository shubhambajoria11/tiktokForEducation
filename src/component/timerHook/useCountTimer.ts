import {useEffect, useState} from 'react';

let interval: NodeJS.Timer | undefined;

const useCountTimer = (
  startingTime: number,
  startTimer: boolean,
): [number, number, boolean, () => void, () => void] => {
  const [timer, setTimer] = useState(startingTime);
  const [isActive, setIsActive] = useState(startTimer);

  function getRemaining(time: number): {
    mins: number;
    secs: number;
  } {
    let mins: number = Math.floor(time / 60);
    let secs: number = time - mins * 60;
    return {mins, secs};
  }

  const {mins, secs} = getRemaining(timer);
  console.log('mins', mins, secs);

  const reset = () => {
    setTimer(startingTime);
    setIsActive(false);
    clearInterval(interval);
  };

  const start = () => {
    setIsActive(true);
  };

  useEffect(() => {
    if (isActive) {
      interval = setInterval(() => {
        setTimer((remainingSec: number) => remainingSec + 1);
      }, 1000);
    } else if (!isActive && timer !== 0) {
      clearInterval(interval);
    }
    return () => clearInterval(interval);
  }, [isActive, timer]);

  useEffect(() => {
    if (mins <= 0 && secs <= 0) {
      setIsActive(false);
    }
  }, [mins, secs]);

  return [mins, secs, isActive, reset, start];
};

export default useCountTimer;
