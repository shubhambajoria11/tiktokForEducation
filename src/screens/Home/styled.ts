import {Dimensions} from 'react-native';
import styled from 'styled-components/native';
import colors from '../../styles/colors';
import {rh, rpx, rw} from '../../styles/Dimen';

const abc = Dimensions.get('screen').height;

export const MainContainer = styled.View`
  background-color: ${colors.backgroundColor};
  flex: 1;
`;
export const HeaderContainer = styled.View`
  background-color: ${colors.darkBlack};
  flex-direction: row;
  justify-content: space-between;
`;

export const HeaderTimerContainer = styled.View`
  flex-direction: row;
  margin: ${rpx(15)}px;
`;

export const TimerText = styled.Text`
  color: ${colors.white};
  opacity: 0.6;
  margin-left: ${rw(3)}px;
`;

export const ClickableContainer = styled.TouchableOpacity``;

export const HeaderFollowTags = styled.View`
  flex-direction: row;
  align-items: center;
`;

export const FollowingText = styled.Text`
  color: ${colors.white};
  opacity: ${(props: {activeFlashCard: any}) => {
    return props.activeFlashCard === 'Following' ? 1 : 0.6;
  }};
  font-weight: ${(props: {activeFlashCard: any}) => {
    return props.activeFlashCard === 'Following' ? 700 : 300;
  }};
`;
export const FollowText = styled.Text`
  color: ${colors.white};
  opacity: ${(props: {activeFlashCard: any}) => {
    return props.activeFlashCard === 'For You' ? 1 : 0.6;
  }};
  font-weight: ${(props: {activeFlashCard: any}) => {
    return props.activeFlashCard === 'For You' ? 700 : 300;
  }};
  margin-left: ${rw(20)}px;
`;

export const HeaderSearch = styled.View`
  align-items: center;
  justify-content: center;
  margin-right: ${rw(20)}px;
`;

export const QuestionsAnswersContainer = styled.TouchableOpacity`
  justify-content: center;
  margin-left: ${rw(16)}px;
  margin-top: ${rh(17)}px;
  margin-right: ${rw(65)}px;
  height: ${rh(500)}px;
`;

export const HorizontalLine = styled.View`
  width: ${rw(286)}px;
  height: 0;
  border-width: ${rw(0.5)}px;
  margin-top: ${rh(24)}px;
  border-color: ${colors.white};
`;

export const QuestionText = styled.Text`
  font-weight: 400;
  font-size: ${rpx(21)}px;
  line-height: ${rh(25)}px;
  align-items: center;
  color: ${colors.white};
`;

export const VerticalCardContainer = styled.View`
  position: absolute;
  bottom: ${rh(52)}px;
  right: ${rw(5)}px;
  justify-content: center;
  align-items: center;
`;

export const AuthorContainer = styled.View``;

export const PlaylistContainer = styled.View`
  position: absolute;
  bottom: 0;
`;

export const AnswerText = styled.Text`
  font-weight: 800;
  font-size: ${rpx(13)}px;
  line-height: ${rh(16)}px;
  color: ${colors.answerText};
  margin-top: ${rh(24)}px;
`;

export const Answer = styled.Text`
  font-weight: 400;
  font-size: ${rpx(21)}px;
  line-height: ${rh(25)}px;
  color: ${colors.white};
  margin-top: ${rh(4)}px;
  opacity: 0.6;
  height: ${rh(225)}px;
  flex-wrap: wrap;
`;

export const HowWellDidYouKnowThis = styled.Text`
  font-weight: 400;
  font-size: ${rpx(15)}px;
  line-height: ${rh(18)}px;
  color: ${colors.white};
  opacity: 0.6;
  margin-top: ${rh(30)}px;
`;

export const ContentContainer = styled.View`
  height: ${abc - rh(140)}px;
  justify-content: space-between;
  justify-content: center;
`;

export const ContentTopContainer = styled.View`
  border-color: ${colors.blue};
`;
export const ContentBottomContainer = styled.View`
  flex: 1;
`;

export const ContentContainerForYou = styled.View`
  height: ${abc - rh(140)}px;
  justify-content: space-between;
  justify-content: center;
`;

export const OptionContainer = styled.View``;
