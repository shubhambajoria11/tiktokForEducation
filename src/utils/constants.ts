export const SERVER_STATUS = {
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const REQUEST_METHODS = Object.freeze({
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  PATCH: 'PATCH',
  DELETE: 'DELETE',
});

export const ASYNC_CONST = {
  accessToken: 'accessToken',
};
