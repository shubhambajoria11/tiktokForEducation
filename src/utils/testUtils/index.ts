export const navigation = {
  reset: jest.fn(),
  navigate: jest.fn(),
  goBack: jest.fn(),
  addListener: jest.fn(),
  isFocused: jest.fn(),
  setParams: jest.fn(),
  dispatch: jest.fn(),
  push: jest.fn(),
  setOptions: jest.fn(),
  getId: jest.fn(),
  getState: jest.fn(),
  canGoBack: jest.fn(),
  getParent: jest.fn(),
  pop: jest.fn(),
  popToTop: jest.fn(),
  removeListener: jest.fn(),
  replace: jest.fn(),
};

export function generateCustomNavigation(
  navigateCallback: jest.Mock<any, any, any>,
  resetCallback?: jest.Mock<any, any, any>,
) {
  const navigationMock = {
    reset: jest.fn(),
    navigate: jest.fn(),
    goBack: jest.fn(),
    addListener: jest.fn(),
    isFocused: jest.fn(),
    setParams: jest.fn(),
    dispatch: jest.fn(),
    push: jest.fn(),
    setOptions: jest.fn(),
    getId: jest.fn(),
    getState: jest.fn(),
    canGoBack: jest.fn(),
    getParent: jest.fn(),
    pop: jest.fn(),
    popToTop: jest.fn(),
    removeListener: jest.fn(),
    replace: jest.fn(),
  };

  navigationMock.navigate = navigateCallback;
  if (resetCallback) navigationMock.reset = resetCallback;
  return navigationMock;
}
