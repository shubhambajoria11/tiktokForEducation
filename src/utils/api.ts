import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {ASYNC_CONST, REQUEST_METHODS} from './constants';

const {POST, GET, PUT, PATCH, DELETE} = REQUEST_METHODS;

const getToken = async () => {
  const token = await AsyncStorage.getItem(ASYNC_CONST.accessToken);
  return token || '';
};

export const api = async (
  url: string,
  {fullUrl = '', method = 'GET', params = {}, headers = {}},
) => {
  const axiosUrl = fullUrl + url;
  const axiosMethod = method.toUpperCase();
  const accessToken = await getToken();
  const token = accessToken;
  const authHeader = {Authorization: `Bearer ${token}`};
  const axiosHeader = {...headers, ...authHeader};
  const axiosRef = axios.create({
    method: axiosMethod,
    url: axiosUrl,
    data: params,
    headers: axiosHeader,
  });

  console.log('axiosUrl', axiosUrl);

  let apiCall: Promise<any>;
  switch (axiosMethod) {
    case POST:
      apiCall = axiosRef.post(axiosUrl, params);
      break;

    case GET:
      apiCall = axiosRef.get(axiosUrl, params);
      break;

    case PUT:
      apiCall = axiosRef.put(axiosUrl, params);
      break;

    case PATCH:
      apiCall = axiosRef.patch(axiosUrl, params);
      break;

    case DELETE:
      apiCall = axiosRef.delete(axiosUrl, {data: params});
      break;

    default:
      apiCall = axiosRef.get(axiosUrl, params);
      break;
  }

  return new Promise((resolve, reject) =>
    apiCall
      .then(response => {
        resolve(response);
      })
      .catch(error => {
        reject(error);
      }),
  );
};

export default api;
