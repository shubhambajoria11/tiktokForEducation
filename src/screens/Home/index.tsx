import React, {useEffect, useState} from 'react';
import Author from '../../component/Author/Author';
import HorizontalBoxes from '../../component/HorizontalBoxes/HorizontalBoxes';
import Playlist from '../../component/playlist/Playlist';
import VerticalCard from '../../component/VericalCard/VerticalCard';
import api from '../../utils/api';
import {REQUEST_METHODS} from '../../utils/constants';
import {
  Answer,
  AnswerText,
  AuthorContainer,
  ClickableContainer,
  ContentBottomContainer,
  ContentContainer,
  ContentContainerForYou,
  ContentTopContainer,
  FollowingText,
  FollowText,
  HeaderContainer,
  HeaderFollowTags,
  HeaderSearch,
  HeaderTimerContainer,
  HorizontalLine,
  HowWellDidYouKnowThis,
  MainContainer,
  PlaylistContainer,
  QuestionsAnswersContainer,
  QuestionText,
  TimerText,
  VerticalCardContainer,
} from './styled';
import SVGIcon from '../../assets/svg/index';
import useCountTimer from '../../component/timerHook/useCountTimer';
import {Alert, FlatList} from 'react-native';
import Options from '../../component/Options/Options';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootTabParamList} from '../../navigation/tabBarNavigation/TabBarNavigator';
import {IFollowing, IForYou} from '../types/types';
import {API_END_POINTS, BASE_URL} from '../../constants/server';
import {
  FOLLOWING,
  FOR_YOU,
  SOMETHING_WENT_WRONG,
} from '../../constants/strings';
import {
  DIGIT_0DOT5,
  DIGIT_1,
  DIGIT_16,
  DIGIT_1DOT5,
  DIGIT_20,
} from '../../constants/numbers';

export type NavigationHomeProps = NativeStackScreenProps<
  RootTabParamList,
  'Home'
>;

interface HomeProps extends NavigationHomeProps {}

const Home: React.FC<HomeProps> = () => {
  const [followingEachData, setFollowingEachData] = useState<IFollowing[]>([]);
  const [forYouEachData, setForYouEachData] = useState<IForYou[]>([]);

  const [activeFlashCard, setActiveFlashCard] = useState<string>(FOLLOWING);

  const [activeClickForShowAnswer, setActiveClickForShowAnswer] =
    useState<boolean>(false);

  const [min, secs] = useCountTimer(DIGIT_1, true);

  const getFollowingQuestionsData = () => {
    api(BASE_URL + API_END_POINTS.following, {
      method: REQUEST_METHODS.GET,
    })
      .then((res: any) => {
        setFollowingEachData((prevData: IFollowing[]) => [
          ...prevData,
          res?.data,
        ]);
      })
      .catch(() => {
        Alert.alert(SOMETHING_WENT_WRONG);
      });
  };

  const getForYouQuestionsData = () => {
    api(BASE_URL + API_END_POINTS.forYou, {
      method: REQUEST_METHODS.GET,
    })
      .then((res: any) => {
        setForYouEachData((prevData: IForYou[]) => [...prevData, res?.data]);
      })
      .catch(() => {
        Alert.alert(SOMETHING_WENT_WRONG);
      });
  };

  useEffect(() => {
    if (activeFlashCard === FOR_YOU) {
      getForYouQuestionsData();
    } else {
      getFollowingQuestionsData();
    }
  }, [activeFlashCard]);

  const renderItem = ({item}: {item: IFollowing | IForYou | any}) => {
    return activeFlashCard === FOLLOWING ? (
      <ContentContainer>
        <ContentTopContainer>
          <QuestionsAnswersContainer
            onPress={() => {
              setActiveClickForShowAnswer(!activeClickForShowAnswer);
            }}>
            <QuestionText>{item.flashcard_front}</QuestionText>
            {activeClickForShowAnswer && <HorizontalLine />}
            {activeClickForShowAnswer && <AnswerText>Answer</AnswerText>}
            {activeClickForShowAnswer && <Answer>{item.flashcard_back}</Answer>}
            {activeClickForShowAnswer && (
              <HowWellDidYouKnowThis>
                How Well Did You Know This
              </HowWellDidYouKnowThis>
            )}
            {activeClickForShowAnswer && <HorizontalBoxes />}
          </QuestionsAnswersContainer>
          <AuthorContainer>
            <Author author={item.user?.name} description={item.description} />
          </AuthorContainer>
          <VerticalCardContainer>
            <VerticalCard />
          </VerticalCardContainer>
        </ContentTopContainer>
        <ContentBottomContainer>
          <PlaylistContainer>
            <Playlist playlistName={item.playlist} />
          </PlaylistContainer>
        </ContentBottomContainer>
      </ContentContainer>
    ) : (
      <ContentContainerForYou>
        <ContentTopContainer>
          <QuestionsAnswersContainer>
            <QuestionText>{item.question}</QuestionText>
            <Options option={item.options} />
          </QuestionsAnswersContainer>
          <AuthorContainer>
            <Author author={item.user?.name} description={item.description} />
          </AuthorContainer>
          <VerticalCardContainer>
            <VerticalCard />
          </VerticalCardContainer>
        </ContentTopContainer>
        <ContentBottomContainer>
          <PlaylistContainer>
            <Playlist playlistName={item.playlist} />
          </PlaylistContainer>
        </ContentBottomContainer>
      </ContentContainerForYou>
    );
  };

  return (
    <MainContainer>
      <HeaderContainer>
        <HeaderTimerContainer>
          <SVGIcon.activity
            width={DIGIT_20}
            height={DIGIT_16}
            strokeWidth={DIGIT_1DOT5}
            fillOpacity={DIGIT_0DOT5}
          />
          <TimerText>
            {min}:{secs}
          </TimerText>
        </HeaderTimerContainer>
        <HeaderFollowTags>
          <ClickableContainer
            activeFlashCard={activeFlashCard}
            onPress={() => {
              setActiveFlashCard(FOLLOWING);
            }}>
            <FollowingText activeFlashCard={activeFlashCard}>
              Following
            </FollowingText>
          </ClickableContainer>
          <ClickableContainer
            activeFlashCard={activeFlashCard}
            onPress={() => {
              setActiveFlashCard(FOR_YOU);
            }}>
            <FollowText activeFlashCard={activeFlashCard}>For You</FollowText>
          </ClickableContainer>
        </HeaderFollowTags>
        <HeaderSearch>
          <SVGIcon.activity
            width={DIGIT_20}
            height={DIGIT_16}
            strokeWidth={DIGIT_1DOT5}
            fillOpacity={DIGIT_0DOT5}
          />
        </HeaderSearch>
      </HeaderContainer>
      {activeFlashCard === FOLLOWING ? (
        <FlatList
          data={followingEachData}
          keyExtractor={(item, index) => item.id.toString() + index.toString()}
          renderItem={renderItem}
          onEndReached={getFollowingQuestionsData}
          onEndReachedThreshold={0.1}
        />
      ) : (
        <FlatList
          data={forYouEachData}
          keyExtractor={(item, index) => item.id.toString() + index.toString()}
          renderItem={renderItem}
          onEndReached={getForYouQuestionsData}
          onEndReachedThreshold={0.1}
        />
      )}
    </MainContainer>
  );
};

export default Home;
