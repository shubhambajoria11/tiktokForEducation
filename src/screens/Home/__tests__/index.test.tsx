import {render, renderHook} from '@testing-library/react-native';
import React from 'react';
import useCountTimer from '../../../component/timerHook/useCountTimer';
import {navigation} from '../../../utils/testUtils';

import Home from '../index';

describe('render Home Screen', () => {
  it('should render correctly', () => {
    const component = render(
      <Home
        navigation={navigation}
        route={{key: '', name: 'Home', params: {}, path: ''}}
      />,
    );
    expect(component).toBeTruthy();
  });
  it('should match with the snapshot', () => {
    const component = render(
      <Home
        navigation={navigation}
        route={{key: '', name: 'Home', params: {}, path: ''}}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('should have the following tab and the for you tab in the header', () => {
    const component = render(
      <Home
        navigation={navigation}
        route={{key: '', name: 'Home', params: {}, path: ''}}
      />,
    );
    expect(component).toBeTruthy();
  });

  it('shoould render the timer functionality', () => {
    render(
      <Home
        navigation={navigation}
        route={{key: '', name: 'Home', params: {}, path: ''}}
      />,
    );

    const startingTime = 1;
    const startTimer = false;
    const {result} = renderHook(() => useCountTimer(startingTime, startTimer));

    expect(result.current[3]).toBeDefined();
  });
});
