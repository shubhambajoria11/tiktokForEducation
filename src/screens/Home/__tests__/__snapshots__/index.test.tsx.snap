// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Home Screen should match with the snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#00425A",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#000000",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "marginBottom": 18.5,
          "marginLeft": 18.5,
          "marginRight": 18.5,
          "marginTop": 18.5,
        }
      }
    >
      <
        fillOpacity={0.5}
        height={16}
        strokeWidth={1.5}
        width={20}
      />
      <Text
        style={
          {
            "color": "#ffffff",
            "marginLeft": 6,
            "opacity": 0.6,
          }
        }
      >
        0
        :
        1
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text
          activeFlashCard="Following"
          style={
            {
              "color": "#ffffff",
              "fontWeight": "700",
              "opacity": 1,
            }
          }
        >
          Following
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text
          activeFlashCard="Following"
          style={
            {
              "color": "#ffffff",
              "fontWeight": "300",
              "marginLeft": 40,
              "opacity": 0.6,
            }
          }
        >
          For You
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "marginRight": 40,
        }
      }
    >
      <
        fillOpacity={0.5}
        height={16}
        strokeWidth={1.5}
        width={20}
      />
    </View>
  </View>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.1}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
